lib/safe-pg-migrations/plugins/helpers/index_definition_comparator.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module SafePgMigrations::Helpers
+  class IndexDefinitionComparator
+    def initialize(index_definition_a, index_definition_b)
+      @index_definition_a = index_definition_a
+      @index_definition_b = index_definition_b
+    end
+
+    def equal?
+      return false unless %i[table name type comment lengths orders opclasses].all? do |attribute|
+        options_equal? @index_definition_a.public_send(attribute), @index_definition_b.public_send(attribute)
+      end
+
+      wheres_equal? && columns_equal? && uniques_equal? && usings_equal?
+    end
+
+    private
+
+    def usings_equal?
+      options_equal?(@index_definition_a.using || 'btree', @index_definition_b.using || 'btree')
+    end
+
+    def uniques_equal?
+      @index_definition_a.unique.presence == @index_definition_b.unique.presence
+    end
+
+    def columns_equal?
+      @index_definition_a.columns.map(&:to_s).sort == @index_definition_b.columns.map(&:to_s).sort
+    end
+
+    def wheres_equal?
+      normalize_where = lambda do |where|
+        where
+          .to_s
+          .gsub(/\s+/, ' ') # replacing any duplicated space with one
+          .gsub(/^\s*\((.*)\)\s*$/, '\1') # removing trailing ()
+          .strip
+      end
```

Comment:
This is clearly *NOT* exhaustive; for example, two conditions being swapped is not handled correctly.

However, I think this is enough, as the end goal is to detect the same migration being executed twice in a row.

Having an error here is possible here, but would require that: 
* the name is the same; 
* a semantically similar definition;
* a different actual definition. 

This can happen if the index is forgotten for example (as it was the case in the bug signaled in TT-7044, the name was already existing with a different definition). In that case, I don't think it's problematic to raise an error.